R/dist.R

```diff
@@ -829,67 +880,515 @@ tune_inv_gamma <- function(lower = 2, upper = 21) {
+#'
+#' @return A delay distribution representing the sum of the two delays
+#' (with class [dist_spec()])
+#'
+#' @author Sebastian Funk
+#' @author Sam Abbott
+#' @importFrom stats convolve
+dist_spec_plus <- function(e1, e2, tolerance = 0.001) {
+  ## process delay distributions
+  delays <- c(e1, e2)
+  ## combine any nonparametric delays that can be combined
+  if (sum(delays$fixed) > 1) {
+    new_pmf <- 1L
+    group_starts <- c(1L, cumsum(delays$np_pmf_length) + 1L)
+    for (i in seq_len(length(group_starts) - 1L)) {
+      new_pmf <- stats::convolve(
+        new_pmf,
+        rev(delays$np_pmf[seq(group_starts[i], group_starts[i + 1L] - 1L)]),
+        type = "open"
+      )
+    }
+    if (!is.infinite(tolerance)) {
+      if (length(new_pmf) > 1) {
+        cdf <- cumsum(new_pmf)
+        new_pmf <- new_pmf[c(TRUE, (1 - cdf[-length(cdf)]) >= tolerance)]
+        new_pmf <- new_pmf / sum(new_pmf)
       }
-      ret$max <- ret$mean_mean
-      if (ret$mean_sd > 0) {
-        stop(
-          "When a distribution has sd == 0 and ",
-          "sd_sd == 0 then mean_sd must be 0, too."
-        )
+    }
+    delays$np_pmf <- new_pmf
+    delays$fixed <- c(1, rep(0, delays$n_p))
+    delays$n_np <- 1
+    delays$n <- delays$n_p + 1
+    delays$np_pmf_max <- length(delays$np_pmf)
+    delays$np_pmf_length <- length(delays$np_pmf)
+  }
+  return(delays)
+}
+
+#' Creates a delay distribution as the sum of two other delay distributions
+#'
+#' This is done via convolution with `stats::convolve()`. Nonparametric delays
+#' that can be combined are processed together, and their cumulative
+#' distribution function is truncated at a specified tolerance level, ensuring
+#' numeric stability.
+#'
+#' @return A delay distribution representing the sum of the two delays
+#' (with class [dist_spec()])
+#' @inheritParams dist_spec_plus
+#' @author Sebastian Funk
+#' @method + dist_spec
+#' @export
+#' @examples
+#' # A fixed lognormal distribution with mean 5 and sd 1.
+#' lognormal <- dist_spec(
+#'   mean = 1.6, sd = 1, max = 20, distribution = "lognormal"
+#' )
+#' lognormal + lognormal
+#'
+#' # An uncertain gamma distribution with mean 3 and sd 2
+#' gamma <- dist_spec(
+#'   mean = 3, sd = 2, mean_sd = 0.5, sd_sd = 0.5, max = 20,
+#'   distribution = "gamma"
+#' )
+#' lognormal + gamma
+#'
+#' # Using tolerance parameter
+#' EpiNow2:::dist_spec_plus(lognormal, lognormal, tolerance = 0.5)
+`+.dist_spec` <- function(e1, e2) {
+  dist_spec_plus(e1, e2, tolerance = 0.001)
+}
+
+#' Combines multiple delay distributions for further processing
+#'
+#' This combines the parameters so that they can be fed as multiple delay
+#' distributions to [epinow()] or [estimate_infections()].
+#'
+#' @param ... The delay distributions (from calls to [dist_spec()]) to combine
+#' @return Combined delay distributions (with class [dist_spec()]`)
+#' @author Sebastian Funk
+#' @method c dist_spec
+#' @importFrom purrr transpose map
+#' @export
```

Comment:
```suggestion
#' @internal
```
so as to avoid confusion with `+` (which is the external-facing function)

Reply:
This prevents users being able to do things like plot(c(lognormal, lognormal + lognormal)) which does seem handy.

Reply:
Dropped in https://github.com/epiforecasts/EpiNow2/pull/363/commits/749dc8e03b0cedbd990406662929bd80c2f80a61